# Basic Data Exploration



In [ ]:
import pandas as pd

In [ ]:
!gdown --id 1BsFTDLymSMx047KaDmLCjOB5Yt2FYpNr

Downloading...
From: https://drive.google.com/uc?id=1BsFTDLymSMx047KaDmLCjOB5Yt2FYpNr
To: /content/melb_data.csv
100% 2.09M/2.09M [00:00<00:00, 143MB/s]


In [ ]:
melbourne_file_path = '/content/melb_data.csv' #data sudah saya upload ke files di Collab
melbourne_data = pd.read_csv(melbourne_file_path) #membaca data sesuai nama variabel
melbourne_data.describe() #print summary data

Rooms         Price  ...    Longtitude  Propertycount
count  13580.000000  1.358000e+04  ...  13580.000000   13580.000000
mean       2.937997  1.075684e+06  ...    144.995216    7454.417378
std        0.955748  6.393107e+05  ...      0.103916    4378.581772
min        1.000000  8.500000e+04  ...    144.431810     249.000000
25%        2.000000  6.500000e+05  ...    144.929600    4380.000000
50%        3.000000  9.030000e+05  ...    145.000100    6555.000000
75%        3.000000  1.330000e+06  ...    145.058305   10331.000000
max       10.000000  9.000000e+06  ...    145.526350   21650.000000

[8 rows x 13 columns]

#Memilih Data Untuk Modelling

In [ ]:
melbourne_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [ ]:
melbourne_data = melbourne_data.dropna(axis=0) #melakukan drop apabila values suatu data ada yang missing

In [ ]:
y = melbourne_data.Price #buat variable y untuk target prediksi

In [ ]:
#memilih parameter kolom yang ingin dijadikan Features 
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude'] 
X = melbourne_data[melbourne_features] #buat variabel penampung List

In [ ]:
X.describe() #melihat isi dari X

Rooms     Bathroom      Landsize    Lattitude   Longtitude
count  6196.000000  6196.000000   6196.000000  6196.000000  6196.000000
mean      2.931407     1.576340    471.006940   -37.807904   144.990201
std       0.971079     0.711362    897.449881     0.075850     0.099165
min       1.000000     1.000000      0.000000   -38.164920   144.542370
25%       2.000000     1.000000    152.000000   -37.855438   144.926198
50%       3.000000     1.000000    373.000000   -37.802250   144.995800
75%       4.000000     2.000000    628.000000   -37.758200   145.052700
max       8.000000     8.000000  37000.000000   -37.457090   145.526350

In [ ]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

In [ ]:
from sklearn.tree import DecisionTreeRegressor

# mendefinisikan model, dengan random state dengan minimal 1 tiap run
melbourne_model = DecisionTreeRegressor(random_state=1)

# untuk fitting model
melbourne_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

In [ ]:
print("membuat prediksi dari 5 rumah di bawah ini:")
print(X.head())
print("Hasil Prediksi Harga")
print(melbourne_model.predict(X.head())) #mengeprint prediction target yaitu y atau Price dari dataset di atas

membuat prediksi dari 5 rumah di bawah ini:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
Hasil Prediksi Harga
[1035000. 1465000. 1600000. 1876000. 1636000.]


#Model Validation

In [ ]:
from sklearn.metrics import mean_absolute_error
#untuk menghitung eror absolut dari rata2 hasil prediksi di atas
predicted_home_prices = melbourne_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

1115.7467183128902

In [ ]:
from sklearn.model_selection import train_test_split

#split data training dan validasi
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 1)
# pilih model "DecisionTreeRegressor"
melbourne_model = DecisionTreeRegressor(random_state = 1)
# model fitting
melbourne_model.fit(train_X, train_y)

# buat prediksi dari data validasi per 5 rumah
val_predictions = melbourne_model.predict(val_X)
ErrorMean = mean_absolute_error(val_y, val_predictions)
print(ErrorMean)

251688.7630729503


#Underfitting and Overfitting

In [ ]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

#buat sebuah function untuk mengatur decision treenya
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [ ]:
#perbandingan max leaf nodes denga Mean Absolute Errornya
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  369673
Max leaf nodes: 50  		 Mean Absolute Error:  266644
Max leaf nodes: 500  		 Mean Absolute Error:  243613
Max leaf nodes: 5000  		 Mean Absolute Error:  256227


Dilihat dari data dia atas maka yang paling optimal ada 500 leaf nodes karena memiliki MAE 2433613 yang mana di model validation memiliki nilai 251688.7630729503 jadi tidak overfitting atau underfitting

#Random Forests

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

190414.59149025998
